src/v/cluster/controller_backend.cc

```diff
@@ -950,51 +1072,147 @@ ss::future<> controller_backend::finish_partition_update(
+    if (!partition) {
+        co_return errc::partition_not_exists;
+    }
     // wait for configuration update, only declare success
     // when configuration was actually updated
     auto update_ec = check_configuration_update(
       _self, partition, replicas, rev);
+
     if (!update_ec) {
-        return ss::make_ready_future<std::error_code>(errc::success);
+        co_return errc::success;
     }
     // we are the leader, update configuration
-    if (partition->is_elected_leader()) {
-        auto brokers = create_brokers_set(replicas, _members_table.local());
-        vlog(
-          clusterlog.debug,
-          "updating partition {} replica set with {}",
-          ntp,
-          replicas);
-
-        auto f = partition->update_replica_set(std::move(brokers), rev);
-        return ss::with_timeout(
-                 model::timeout_clock::now() + std::chrono::seconds(5),
-                 std::move(f))
-          .then_wrapped([](ss::future<std::error_code> f) {
-              try {
-                  return f.get0();
-              } catch (const ss::timed_out_error& e) {
-                  return make_error_code(errc::timeout);
-              }
-          })
-          .then([this, partition, replicas, rev](std::error_code) {
-              return check_configuration_update(
-                _self, partition, replicas, rev);
-          });
+    if (partition->is_leader()) {
+        auto f = func(partition);
+        try {
+            // TODO: use configurable timeout here
+            auto err = co_await ss::with_timeout(
+              model::timeout_clock::now() + std::chrono::seconds(5),
+              std::move(f));
+            if (err) {
+                co_return err;
+            }
+        } catch (const ss::timed_out_error& e) {
+            co_return make_error_code(errc::timeout);
+        }
+        co_return check_configuration_update(_self, partition, replicas, rev);
     }
 
-    return ss::make_ready_future<std::error_code>(errc::not_leader);
+    co_return errc::not_leader;
+}
+
+ss::future<std::error_code> controller_backend::cancel_replica_set_update(
+  const model::ntp& ntp,
+  const std::vector<model::broker_shard>& replicas,
+  model::revision_id rev) {
+    /**
+     * Following scenarios can happen in here:
+     * - node is a leader for current partition => cancel raft group
+     * reconfiguration
+     * - node is not a leader for current partition => check if config is
+     *   equal to requested, if not return failure
+     */
+
+    return apply_configuration_change_on_leader(
+      ntp,
+      replicas,
+      rev,
+      [this, &ntp, rev, replicas](ss::lw_shared_ptr<partition> p) {
+          const auto raft_cfg_update_finished
+            = are_configuration_replicas_up_to_date(
+              p->group_configuration(), replicas);
+
+          // raft already finished its part, we need to move replica back
+          if (raft_cfg_update_finished) {
```

Comment:
exactly, we do not synchronize both, in current implementation cancelling partition depends on the state in `topic_table` i.e. if update is still marked in progress it may be cancelled, here we need to make sure that the partition replicas are consistent with the state in topics table.